src/pops.cpp

```diff
@@ -124,177 +125,145 @@ List pops_model(int random_seed,
-    output_frequency = time_step;
-  }
-  
-  int first_mortality_year = mortality_time_lag;
-  
+
   std::vector<IntegerMatrix> infected_vector;
   std::vector<IntegerMatrix> susceptible_vector;
   std::vector<IntegerMatrix> mortality_tracker_vector;
   std::vector<IntegerMatrix> mortality_vector;
   std::vector<IntegerMatrix> resistant_vector;
   std::vector<IntegerMatrix> total_host_vector;
   std::vector<IntegerMatrix> dispersers_vector;
-  
-  StepUnit step_unit = step_unit_enum_from_string(time_step);
 
-  // Define simulation time step
-  Scheduler scheduler(dd_start, dd_end, step_unit, 1);
-  // Define spread schedule
-  std::vector<bool> spread_schedule = scheduler.schedule_spread(season);
-  // Define spread rate schedule
-  std::vector<bool> spread_rate_schedule = scheduler.schedule_action_end_of_year();
-  // Define mortality schedule
-  std::vector<bool> mortality_schedule = scheduler.schedule_action_end_of_year();
-  // Define lethality schedule
-  std::vector<bool> lethality_schedule = scheduler.schedule_action_yearly(lethal_temperature_month, 1);
-  // Define output schedule
-  std::vector<bool> output_schedule;
-  if (output_frequency == "year") {
-    output_schedule = scheduler.schedule_action_end_of_year();
-  } else if (output_frequency == "month") {
-    output_schedule = scheduler.schedule_action_monthly();
-  } else if (output_frequency == "week") {
-    if (time_step == "day") {
-      output_schedule = scheduler.schedule_action_nsteps(7);
-    } else if (time_step == "week") {
-      output_schedule = scheduler.schedule_action_nsteps(1);
-    }
-  } else if (output_frequency == "day") {
-    output_schedule = scheduler.schedule_action_nsteps(1);
-  }
-  
-  Treatments<IntegerMatrix, NumericMatrix> treatments(scheduler);
-  bool use_treatments = false;
+  config.set_step_unit(time_step);
+
+  config.create_schedules();
+
+  Treatments<IntegerMatrix, NumericMatrix> treatments(config.scheduler());
   for (unsigned t = 0; t < treatment_maps.size(); t++) {
     treatments.add_treatment(treatment_maps[t], pops::Date(treatment_dates[t]), pesticide_duration[t], treatment_application);
-    use_treatments = true;
+    config.use_treatments = true;
   }
-  
-  unsigned count_lethal = get_number_of_scheduled_actions(lethality_schedule);
-    if (use_lethal_temperature && count_lethal > temperature.size()){
-      Rcerr << "Not enough years of temperature data" << std::endl;
-    }
-    
-  unsigned count_weather = get_number_of_scheduled_actions(spread_schedule);
-      if (weather && count_weather > weather_coefficient.size()) {
-        Rcerr << "Not enough indices of weather coefficient data" << std::endl;
-      }
-  
-  unsigned mortality_simulation_year;
-  
-  unsigned spread_rate_outputs = get_number_of_scheduled_actions(spread_rate_schedule);
-  SpreadRate<IntegerMatrix> spreadrate(infected, ew_res, ns_res, spread_rate_outputs);
-  // Define movement schedule  
+
+  unsigned count_lethal = config.num_lethal();
```

Comment:
num_lethal() can be only called when lethal temperature is in use:

```
unsigned num_lethal()
    {
        if (!use_lethal_temperature)
            throw std::logic_error(
                "num_lethal() not available when use_lethal_temperature is false");
        if (!schedules_created_)
            throw std::logic_error(
                "Schedules were not created before calling num_lethal()");
        return get_number_of_scheduled_actions(lethal_schedule_);
    }
```

Reply:
<img width="494" alt="Screen Shot 2020-07-16 at 11 42 40 AM" src="https://user-images.githubusercontent.com/63056769/87692287-92f88a00-c759-11ea-8933-56a64f45f989.png">